File: python/83_beautiful_strings.py
import re
import sys

with open(sys.argv[1], 'r') as fh:
    for line in fh.readlines():
        line = line.strip()
        if line == '':
            continue

        line = re.sub(r'[^a-z]', '', line.lower())

        alphabet = frozenset(list(line))
        counts = sorted([len(re.findall(l, line)) for l in alphabet], reverse = True)
        score = 0
        for i in range(len(counts)):
            score += (26 - i) * counts[i]

        print score

File: perl/83_beautiful_strings.pl
open my $fh, '<', $ARGV[0];

for my $line ( <$fh> ) {
	$line =~ s/[\n\r]+//g;
	next if $line eq '';

	$line = lc $line;
	$line =~ s/[^a-z]//g;

	my %alphabet = map { $_ => $_ } split //, $line;
	my @counts = ();
	for my $l ( keys %alphabet ) {
		my $c =()= $line =~ m/$l/g;
		push @counts, $c;
	}

	@counts = sort { $b <=> $a } @counts;
	my $score = 0;
	for ( my $i = 0; $i != scalar @counts; $i++ ) {
		$score += (26 - $i) * $counts[$i];
	}

	print "$score\n";
}

close $fh;

exit 0;

